pkg/cloud/ratelimit.go

```diff
@@ -64,25 +66,54 @@ type AcceptRateLimiter struct {
 }
 
 // Accept wraps an Acceptor and blocks on Accept or context.Done(). Key is ignored.
-func (rl *AcceptRateLimiter) Accept(ctx context.Context, key *RateLimitKey) error {
-	ch := make(chan struct{})
-	go func() {
+func (rl *AcceptRateLimiter) Accept(ctx context.Context, _ *RateLimitKey) error {
+	return acceptInternal(ctx, func() {
 		rl.Acceptor.Accept()
-		close(ch)
-	}()
-	select {
-	case <-ch:
-		break
-	case <-ctx.Done():
-		return ctx.Err()
-	}
-	return nil
+	})
 }
 
 // Observe does nothing.
 func (rl *AcceptRateLimiter) Observe(_ error, _ *RateLimitKey) {
 }
 
+// ThrottlingStrategy handles delays based on feedbacks provided.
+type ThrottlingStrategy interface {
+	// GetDelay returns the delay for next request.
+	GetDelay() time.Duration
+	// PushFeedback recalculates the next delay based on the feedback.
+	PushFeedback(err error)
+}
+
+// StrategyRateLimiter wraps a ThrottlingStrategy with RateLimiter parameters.
+type StrategyRateLimiter struct {
+	// strategy is the underlying throttling strategy.
+	strategy ThrottlingStrategy
+	lock     sync.Mutex
+	clock    clock.Clock
+}
+
+// NewStrategyRateLimiter returns a StrategyRateLimiter backed by the provided ThrottlingStrategy.
+func NewStrategyRateLimiter(strategy ThrottlingStrategy) *StrategyRateLimiter {
+	return &StrategyRateLimiter{
+		strategy: strategy,
+		clock:    clock.RealClock{},
+	}
+}
+
+// Accept block for the delay provided by the strategy or until context.Done(). Key is ignored.
+func (rl *StrategyRateLimiter) Accept(ctx context.Context, _ *RateLimitKey) error {
+	return acceptInternal(ctx, func() {
+		rl.lock.Lock()
```

Comment:
Yes, it doesn't make sense, since it'll be blocked anyway, but by delay. Having an explicit lock makes it easier to track required delay for each request, without the lock it should be calculated like some base delay from the strategy + remaining delay from the request(s) in the queue.